internal/httpapi/httpapi.go

```diff
@@ -96,133 +101,184 @@ func NewServer(r status.Reporter, em emitter.Emitter) (*Server, error) {
-
-	server := &http.Server{
-		Handler: router,
-		Addr:    s.Ingest.address,
-	}
+func (s *Server) serveIngest(_ context.Context) error {
+	serverErr := make(chan error, 1)
+
+	go func() {
+		defer close(serverErr)
+		s.logger.WithFields(logrus.Fields{
+			"address": s.Ingest.address,
+		}).Debug("Ingest API starting listening.")
+
+		router := httprouter.New()
+		router.GET(ingestAPIPathReady, s.handleReady)
+		router.POST(ingestAPIPath, s.handleIngest)
+
+		server := &http.Server{
+			Handler: router,
+			Addr:    s.Ingest.address,
+		}
+
+		if s.Ingest.tls.enabled {
+			if s.Ingest.tls.validateClient {
+				err := addMTLS(server, s.Ingest.tls.caPath)
+				if err != nil {
+					serverErr <- fmt.Errorf("creating mTLS server: %w", err)
+
+					return
+				}
+			}
 
-	if s.Ingest.tls.enabled {
-		if s.Ingest.tls.validateClient {
-			err := addMTLS(server, s.Ingest.tls.caPath)
+			s.logger.Debug("starting tls server")
+			err := server.ListenAndServeTLS(s.Ingest.tls.certPath, s.Ingest.tls.keyPath)
 			if err != nil {
-				return fmt.Errorf("creating mTLS server: %w", err)
+				serverErr <- fmt.Errorf("starting tls server: %w", err)
+
+				return
 			}
 		}
 
-		s.logger.Debug("starting tls server")
-		err := server.ListenAndServeTLS(s.Ingest.tls.certPath, s.Ingest.tls.keyPath)
+		err := server.ListenAndServe()
 		if err != nil {
-			return fmt.Errorf("starting tls server: %w", err)
+			s.logger.WithError(err).Error("Ingest server error")
 		}
-	}
+		serverErr <- err
+	}()
 
-	err := server.ListenAndServe()
-	if err != nil {
-		return fmt.Errorf("starting Ingest server: %w", err)
-	}
-
-	return ErrServerStopped
+	return s.waitUntilReadyOrError(s.Ingest.address, ingestAPIPathReady, s.Ingest.tls.enabled, s.Status.tls.validateClient, serverErr)
 }
 
-func (s *Server) allReadyOrDisabled(ingestReady, statusReady bool) bool {
-	if s.Ingest.enabled && !ingestReady {
-		return false
+// waitUntilReadyOrError makes http request to address if server didn't return an error.
+// It has a timeout to prevent staying in an infinite loop
+func (s *Server) waitUntilReadyOrError(address string, path string, tlsEnabled bool, validateTLSClient bool, serverErrCh <-chan error) error {
+	var err error
+	client := http.Client{}
+	scheme := "http"
+
+	if tlsEnabled {
+		scheme = "https"
+		client.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true, //nolint:gosec
+			},
+		}
 	}
-	if s.Status.enabled && !statusReady {
-		return false
+
+	url := fmt.Sprintf("%s://%s%s", scheme, address, path)
+	timer := time.NewTimer(readinessProbeTimeout)
+LOOP:
+	for {
+		// we don't test the local server when tls is enabled and validate client is false
+		if (tlsEnabled && !validateTLSClient) || s.isGetSuccessful(client, url) {
+			break
+		}
+		// if the server is not running and returned an error we stop trying and return the error
+		select {
+		case err = <-serverErrCh:
+			if err != nil {
+				break LOOP
```

Comment:
What do you think of replacing this with `return err`?

Reply:
totally agree, not sure what happened there.... :homer-dissapearing: ...
fixed in https://github.com/newrelic/infrastructure-agent/pull/1438/commits/06ab4a35b937be4ac31e7da8a72361d858d39b6c